event/src/unvalidated/builder.rs

```diff
@@ -1,343 +1,324 @@
+    pub fn with_should_index(mut self, should_index: bool) -> Self {
+        self.state.should_index = Some(should_index);
         self
     }
-}
 
-impl<D> Sep for InitWithDataBuilder<D> {
-    fn with_sep(mut self, sep: String) -> Self {
-        self.sep = Some(sep);
+    /// Specify the data.
+    pub fn with_data(mut self, data: D) -> Self {
+        self.state.data = Some(data);
         self
     }
+    /// Build the event.
+    pub fn build(self) -> unvalidated::init::Payload<D> {
+        let header = unvalidated::init::Header::new(
+            vec![self.state.controller],
+            self.state.sep.key,
+            self.state.sep.value,
+            self.state.should_index,
+            self.state.unique,
+        );
+        unvalidated::init::Payload::new(header, self.state.data)
+    }
+}
+
+/// Builder for constructing an [`unvalidated::data::Payload`].
+#[derive(Default)]
+#[allow(private_bounds)]
+pub struct DataBuilder<S: unvalidated::builder::DataBuilderState> {
+    state: S,
 }
 
-impl<D> InitWithDataBuilder<D> {
-    /// Build unsigned init event with data
-    pub async fn build(self) -> anyhow::Result<init::Payload<D>> {
-        if self.controllers.is_empty() {
-            anyhow::bail!("controllers must not be empty");
+/// State of the builder
+pub trait DataBuilderState {}
+
+/// Initial state
+pub struct DataBuilderEmpty;
+impl DataBuilderState for DataBuilderEmpty {}
+
+/// State with id added
+pub struct DataBuilderWithId {
+    id: Cid,
+}
+impl DataBuilderState for DataBuilderWithId {}
+impl DataBuilder<DataBuilderEmpty> {
+    /// Specify the id.
+    pub fn with_id(self, id: Cid) -> DataBuilder<DataBuilderWithId> {
+        DataBuilder {
+            state: DataBuilderWithId { id },
         }
-        let sep = self
-            .sep
-            .ok_or_else(|| anyhow::anyhow!("sep must be specified"))?;
-        let header = init::Header::new(self.controllers, sep, self.additional);
-        let payload = init::Payload::new(header, Some(self.data));
-        Ok(payload)
     }
 }
 
-/// Builder for data event
-pub struct DataBuilder<T> {
+/// State with prev added
+pub struct DataBuilderWithPrev {
     id: Cid,
     prev: Cid,
-    data: T,
-    sep: Option<String>,
-    additional: ValueMap,
+}
+impl DataBuilderState for DataBuilderWithPrev {}
+impl DataBuilder<DataBuilderWithId> {
+    /// Specify the prev.
+    pub fn with_prev(self, prev: Cid) -> DataBuilder<DataBuilderWithPrev> {
+        DataBuilder {
+            state: DataBuilderWithPrev {
+                id: self.state.id,
+                prev,
+            },
+        }
+    }
 }
 
-impl<T> Additional for DataBuilder<T> {
-    fn with_additional(mut self, key: String, value: Value) -> Self {
-        self.additional.insert(key, value);
-        self
+/// State with data added, also supports the optional should_index field.
+pub struct DataBuilderWithData<D> {
+    id: Cid,
+    prev: Cid,
+    data: D,
+    should_index: Option<bool>,
```

Comment:
we should add `with_sep` to data builder too, right? we need sep/sepValue to build a recon event ID.

Reply:
Data events don't have the separator in their header.  So the separator needs to be fetched from the init event to build the event ID.

Reply:
oh right. I knew that.. and still think it's annoying 😅